src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java

```diff
@@ -1264,6 +1267,41 @@ public void accessPermittedOnlyFromCurrentBuild() throws Throwable {
             r.buildAndAssertSuccess(main);
         });
     }
+
+    @Ignore("TODO safe fix still TBD")
+    @Test public void placeholderTaskInQueueButAssociatedBuildComplete() throws Throwable {
+        Path tempQueueFile = tmp.newFile().toPath();
+        sessions.then(r -> {
+            WorkflowJob p = r.createProject(WorkflowJob.class, "p");
+            p.setDefinition(new CpsFlowDefinition("node('custom-label') { }", true));
+            WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+            // Get into a state where a PlaceholderTask is in the queue.
+            while (true) {
+                Queue.Item[] items = Queue.getInstance().getItems();
+                if (items.length == 1 && items[0].task instanceof ExecutorStepExecution.PlaceholderTask) {
+                    break;
+                }
+                Thread.sleep(500L);
+            }
+            // Copy queue.xml to a temp file while the PlaceholderTask is in the queue.
+            r.jenkins.getQueue().save();
+            Files.copy(sessions.getHome().toPath().resolve("queue.xml"), tempQueueFile, StandardCopyOption.REPLACE_EXISTING);
+            // Create a node with the correct label and let the build complete.
+            DumbSlave node = r.createOnlineSlave(Label.get("custom-label"));
+            r.assertBuildStatusSuccess(r.waitForCompletion(b));
+        });
+        // Copy the temp queue.xml over the real one. The associated build has already completed, so the queue now
+        // has a bogus PlaceholderTask.
+        Files.copy(tempQueueFile, sessions.getHome().toPath().resolve("queue.xml"), StandardCopyOption.REPLACE_EXISTING);
+        sessions.then(r -> {
+            WorkflowJob p = r.jenkins.getItemByFullName("p", WorkflowJob.class);
+            WorkflowRun b = p.getBuildByNumber(1);
+            assertFalse(b.isLogUpdated());
+            r.assertBuildStatusSuccess(b);
+            assertThat(Queue.getInstance().getItems(), emptyArray()); // This assertion fails.
```

Comment:
Turns out the test as-is is flaky right now. If you add a 5 second sleep before this line so that the agent has time to come online, then the placeholder task attempts to run, fails, and the assertion succeeds.

Specifically [this call to `StepContext.get`](https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/f832bc1ac7457afe18fb6d62b199c7f6c29335cf/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java#L802) fails with this exception because it expects the execution to still be in progress when it has already completed:

```
java.lang.IllegalStateException: completed or broken execution
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.onLoad(CpsFlowExecution.java:764)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.getExecution(WorkflowRun.java:691)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.get(WorkflowRun.java:963)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.getExecution(CpsStepContext.java:217)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.getThreadGroupSynchronously(CpsStepContext.java:242)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.access$000(CpsStepContext.java:97)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext$1.call(CpsStepContext.java:263)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext$1.call(CpsStepContext.java:261)
	at jenkins.util.ContextResettingExecutorService$2.call(ContextResettingExecutorService.java:46)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused: java.util.concurrent.ExecutionException
	at com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:289)
	at com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:276)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:111)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.getThreadGroupSynchronously(CpsStepContext.java:247)
Caused: java.io.IOException
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.getThreadGroupSynchronously(CpsStepContext.java:249)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.getThreadSynchronously(CpsStepContext.java:236)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.doGet(CpsStepContext.java:293)
	at org.jenkinsci.plugins.workflow.support.DefaultStepContext.get(DefaultStepContext.java:75)
	at org.jenkinsci.plugins.workflow.support.steps.ExecutorStepExecution$PlaceholderTask$PlaceholderExecutable.run(ExecutorStepExecution.java:806)
	at hudson.model.ResourceController.execute(ResourceController.java:99)
```

Somewhat interestingly, this does not trigger the "already completed" log messages from `CpsStepContext.completed` because the copy of `CpsStepContext` in question was serialized before the step completed and so it thinks it is completing for the first time. This means that no warnings are logged at all when this happens, we just save the above exception into `CpsStepContext.outcome` and then `CpsStepContext.scheduleNextRun` fails silently in the `onFailure` arm of the future passed to `runInCpsVmThread`.

I can de-flake the test by removing the node before the restart, which I think is still somewhat analogous to what might happen with a non-static agent after a restart, but I'm not sure.

Reply:
> I can de-flake the test by removing the node before the restart, which I think is still somewhat analogous to what might happen with a non-static agent after a restart

:+1:

Reply:
De-flaked in 00cdd69.